web/reader.js

```diff
@@ -16,8 +16,101 @@ var PDFNavigator = cozy.Control.Navigator.extend({
   EOT: true
 });
 
+var PDFSearch = cozy.Control.Search.extend({
+
+  onAdd: function(reader) {
+    var self = this;
+    console.log("AHOY AHOY", this);
+    this._reader.on('ready', function() {
+      // updatefindcontrolstate -- emites state=FOUND before all the results are found
+      this._reader.PDFViewerApplication.eventBus.on('updatefindmatchescount', this._waitForResults.bind(this));
+      this._reader.PDFViewerApplication.eventBus.on('updatefindcontrolstate', this._waitForQuery.bind(this));
+    }.bind(this));
+
+    return cozy.Control.Search.prototype.onAdd.apply(this, arguments);
+  },
+
+  submitQuery: function() {
+    this._doWaitForResults = false;
+    this._reader.PDFViewerApplication.findController.executeCommand('find', { source: this._reader, type: 'find', query: this.searchString, phraseSearch: true, caseSensitive: false, entireWord: false, highlightAll: true, findPrevious: null })
+  },
+
+  _waitForQuery: function(argv) {
+    // if ( this.queryTimer ) { clearTimeout(this.queryTimer); }
+    if ( argv.state == 0 || argv.state == 2 ) {
+      // the query is done, so we can start computing the results
+      this._doWaitForResults = true;
+    } else if ( argv.state == 1 ) {
+      this._data = { highlight_off: "yes", search_results: [] };
+      this._buildResults();
+    }
+    console.log("AHOY updatefindcontrolstate", argv);
+  },
+
+  _waitForResults: function(argv) {
+    if ( ! this._doWaitForResults ) { return; }
+    if ( this.searchTimer ) { clearTimeout(this.searchTimer); }
+    this.searchTimer = setTimeout(this._gatherResults.bind(this), 500);
+    console.log("AHOY updatefindmatchescount", argv.matchesCount.total);
+  },
+
+  _gatherResults: function() {
+    var self = this;
+
+    this._data = { search_results: [], highlight_off: 'yes' };
+    var pdfViewer = this._reader.pdfViewer;
+    var findController = this._reader.PDFViewerApplication.findController;
+
+    for(var i = 0; i < pdfViewer.pagesCount; i++) {
+      var matches = findController._pageMatches[i];
+      if ( ! matches ) { continue ; }
+      var text = findController._pageContents[i];
+      matches.forEach(function(match, index) {
```

Comment:
PDFViewerApplication doesn't do snippets; we try to build snippets from the raw text. Using the HTML from the text layer would be best, but they're lazily loaded.